pkg/network/kuryr_bootstrap.go

```diff
@@ -0,0 +1,620 @@
+	}
+	log.Printf("Pod subnetpool %s present", podSubnetpoolId)
+
+	workerSubnetId, err := findSubnetId(client, "nodes", tag)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to find worker nodes subnet")
+	}
+	log.Printf("Found worker nodes subnet %s", workerSubnetId)
+	routerId, err := findRouterId(client, "openshift-external-router", tag)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to find worker nodes router")
+	}
+	log.Printf("Found worker nodes router %s", routerId)
+	ps, err := getRouterPorts(client, routerId)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed list ports of worker nodes router")
+	}
+
+	if !lookupPort(ps, svcSubnetId) {
+		log.Printf("Ensuring service subnet router port with %s IP", ipStr)
+		portId, err := ensurePort(client, "kuryr-service-subnet-router-port", tag,
+			svcNetId, svcSubnetId, ipStr)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to create service subnet router port")
+		}
+		log.Printf("Service subnet router port %s present, adding it as interface.", portId)
+		err = ensureRouterInterface(client, routerId, nil, &portId)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to create service subnet router interface")
+		}
+	}
+
+	masterSgId, err := findSgId(client, "master", tag)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to find master nodes security group")
+	}
+	log.Printf("Found master nodes security group %s", masterSgId)
+	workerSgId, err := findSgId(client, "worker", tag)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to find worker nodes security group")
+	}
+	log.Printf("Found worker nodes security group %s", workerSgId)
+
+	log.Print("Ensuring pods security group")
+	podSgId, err := ensureSg(client, "kuryr-pods-security-group", tag)
+	log.Printf("Pods security group %s present", podSgId)
+
+	log.Print("Allowing traffic from masters and nodes to pods")
+	err = ensureSgRule(client, podSgId, masterSgId)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to add rule opening traffic from masters")
+	}
+	err = ensureSgRule(client, podSgId, workerSgId)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to add rule opening traffic from workers")
+	}
+	err = ensureSgRule(client, masterSgId, podSgId)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to add rule opening traffic to masters")
+	}
+	err = ensureSgRule(client, workerSgId, podSgId)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to add rule opening traffic to workers")
+	}
+	log.Print("All requried traffic allowed")
+
+	lbClient, err := openstack.NewLoadBalancerV2(provider, gophercloud.EndpointOpts{
+		Region: region,
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create Octavia client")
+	}
+
+	// We need first usable IP from services CIDR
+	// This will get us the first one (subnet IP)
+	ip = svcNet.IP
+	// And this second one (first usable)
+	ip[3] |= 0x01
```

Comment:
Interesting. Why is the openshift API service special? Is this just needed to bootstrap the kuryr-controller?

Reply:
Yes, it's needed for the kuryr-controller. We agreed to check out if we can use the openshift-sdn approach and let kuryr-controller configure that LB on its own.